kernel/src/path.rs

```diff
@@ -1,136 +1,178 @@
-
-            if !is_commit && suffix.starts_with("checkpoint.") {
-                // check if name is just [version].checkpoint.parquet, i.e. we have a classic checkpoint
-                let rest = &suffix[11..]; // strip off the "checkpoint." which is 11 chars
-                is_checkpoint = rest == "parquet";
-                if !is_checkpoint {
-                    // test if we're a multipart checkpoint
-                    let mut split = rest.splitn(3, '.');
-                    let (checkpoint_index, checkpoint_max, ext) = (
-                        get_version_opt(split.next(), MULTIPART_PART_LEN),
-                        get_version_opt(split.next(), MULTIPART_PART_LEN),
-                        split.next(),
-                    );
-                    is_checkpoint = checkpoint_index.is_some()
-                        && checkpoint_max.is_some()
-                        && ext == Some("parquet");
-                }
+impl<Location: AsUrl> ParsedLogPath<Location> {
+    // NOTE: We can't actually impl TryFrom because Option<T> is a foreign struct even if T is local.
+    pub fn try_from(location: Location) -> DeltaResult<Option<ParsedLogPath<Location>>> {
+        let url = location.as_url();
+        let filename = url
+            .path_segments()
+            .ok_or_else(|| Error::invalid_log_path(url))?
+            .last()
+            .unwrap() // "the iterator always contains at least one string (which may be empty)"
+            .to_string();
+        if filename.is_empty() {
+            return Err(Error::invalid_log_path(url));
+        }
+
+        let mut split = filename.split('.');
+
+        // NOTE: str::split always returns at least one item, even for the empty string.
+        let version = split.next().unwrap();
+
+        // Every valid log path starts with a numeric version part. If version parsing fails, it
+        // must not be a log path and we simply return None. However, it is an error if version
+        // parsing succeeds for a wrong-length numeric string.
+        let version = match version.parse().ok() {
+            Some(v) if version.len() == VERSION_LEN => v,
+            Some(_) => return Err(Error::invalid_log_path(url)),
+            None => return Ok(None),
+        };
+
+        // Every valid log path has a file extension as its last part. Return None if it's missing.
+        let split: Vec<_> = split.collect();
+        let extension = match split.last() {
+            Some(extension) => extension.to_string(),
+            None => return Ok(None),
+        };
+
+        // Parse the file type, based on the number of remaining parts
+        let file_type = match split.len() {
+            // Commit file: <n>.json
+            1 if extension == "json" => LogPathFileType::Commit,
+
+            // Single-part checkpoint: <n>.checkpoint.parquet
+            2 if split[0] == "checkpoint" && extension == "parquet" => {
+                LogPathFileType::SinglePartCheckpoint
+            }
```

Comment:
I think we could also do `split.as_slice()`? (just for another idea if that helps readability)
```suggestion
        let file_type = match split.as_slice() {
            // Commit file: <n>.json
            [_] if extension == "json" => LogPathFileType::Commit,

            // Single-part checkpoint: <n>.checkpoint.parquet
            ["checkpoint", _] && extension == "parquet" => {
                LogPathFileType::SinglePartCheckpoint
            }
            ...
```

Reply:
also maybe doing pop_back for the extension would help with clarity so it isn't still in `split`? (would require VecDeque though..)

Reply:
Ohh, that's nice!